rasterio/_vsiopener.pyx

```diff
@@ -0,0 +1,154 @@
+# cython: language_level=3, boundscheck=False
+# distutils: language = c++
+"""Bridge between Python file openers and GDAL VSI.
+
+Based on _filepath.pyx.
+"""
+
+include "gdal.pxi"
+
+import contextlib
+import logging
+from uuid import uuid4
+
+from libc.string cimport memcpy
+
+log = logging.getLogger(__name__)
+
+gdal33_version_checked = False
+gdal33_version_met = False
+
+
+# NOTE: This has to be defined outside of gdal.pxi or other C extensions will
+# try to compile C++ only code included in this header.
+cdef extern from "cpl_vsi_virtual.h":
+    cdef cppclass VSIFileManager:
+        @staticmethod
+        void* GetHandler(const char*)
+
+
+# Prefix for all in-memory paths used by GDAL's VSI system
+# Except for errors and log messages this shouldn't really be seen by the user
+cdef str PREFIX = "/vsipyopener/"
+cdef bytes PREFIX_BYTES = PREFIX.encode("utf-8")
+
+# This is global state for the Python filesystem plugin. It currently only
+# contains path -> PyOpenerBase (or subclass) instances. This is used by
+# the plugin to determine what "files" exist on "disk".
+# Currently the only way to "create" a file in the filesystem is to add
+# an entry to this dictionary. GDAL will then Open the path later.
+cdef _OPENER_REGISTRY = {}
+cdef _OPEN_FILE_OBJS = set()
+
+
+cdef int install_pyopener_plugin(VSIFilesystemPluginCallbacksStruct *callbacks_struct):
+    """Install handlers for python file openers if it isn't already installed."""
+    callbacks_struct = VSIAllocFilesystemPluginCallbacksStruct()
+    callbacks_struct.open = <VSIFilesystemPluginOpenCallback>pyopener_open
+    callbacks_struct.tell = <VSIFilesystemPluginTellCallback>pyopener_tell
+    callbacks_struct.seek = <VSIFilesystemPluginSeekCallback>pyopener_seek
+    callbacks_struct.read = <VSIFilesystemPluginReadCallback>pyopener_read
+    callbacks_struct.close = <VSIFilesystemPluginCloseCallback>pyopener_close
+    callbacks_struct.pUserData = <void*>_OPENER_REGISTRY
+
+    if VSIFileManager.GetHandler("") == VSIFileManager.GetHandler(PREFIX_BYTES):
```

Comment:
You could call VSIGetFileSystemsPrefixes() (return value to be freed with CSLDestroy()) and see if PREFIX_BYTES is already listed, to avoid calling a C++ GDAL method